Function unicode compatiblity

AadityaNair opened the pull request:
Make undefined functions accept unicoded names. Fixes #8960.

AadityaNair committed:
`UndefinedFunction` now accepts Unicoded names. Fixes #8960.

Previously, creating a function with unicode name resulted in an error.

AadityaNair committed:
Added tests for Unicode Compatiblity.

Corrected incompatiblity with python3

the initial naive implementation created a `byte` object in python3.
I converted the byte object to `str` object.

Removed Trailing whitespaces.

AadityaNair commented on sympy/core/function.py:
```diff
@@ -707,6 +707,7 @@ class UndefinedFunction(FunctionClass):
     The (meta)class of undefined functions.
     """
     def __new__(mcl, name, **kwargs):
+        name=name.encode('ascii') # if user provides an UNICODE name
```
`name.encode('ascii', 'ignore')` ignores error generated if non-ascii characters are found.
I thought it might be better, if it raises an error if it encounters such characters than ignore it.

asmeurer replied:
Yes, that is definitely better.

asmeurer commented:
This breaks Python 3, where Unicode names work just fine.

asmeurer commented:
That is, actual Unicode characters, like α.

AadityaNair commented:
@asmeurer What should we do about such characters. At present they would raise an error.

sunny94 commented:
If these Unicode characters, like α are considered then it won't be possible for any Python class to be ASCII string. For example  consider following error
In [52]: Symbol(u'α')
Out[52]: <repr(<sympy.core.symbol.Symbol at 0x7fc71738db40>) failed: UnicodeEncodeError: 'ascii' codec can't encode character u'\u03b1' in position 0: ordinal not in range(128)>

AadityaNair commented:
I have updated the PR. Con you explain the error in https://travis-ci.org/sympy/sympy/jobs/50863162
and why is it coming in that one test ?

asmeurer commented:
You can't use the `u` prefix for strings because we have to support Python 3.2. Use the `u()` function from `sympy.core.compatibility` instead.

asmeurer commented:
I am -1 to this as it stands because it breaks `Function('α')` in Python 3.

AadityaNair committed:
Fixed tests to support python3.2

Tests for python3.2 failed because `Function(u'k')` raised a `SyntaxError`.
Fixed by using `u()` from `sympy.core.compatiblity`.

AadityaNair commented:
About supporting ***α***, Can we create a mapping between these characters and the `greeks` set somewhere in `alphabets.py` and use them.

AadityaNair committed:
`Function()` now accepts greek characters as function name.

Removed extra Whitespaces.

AadityaNair commented:
@asmeurer Updated the PR. `Function()` now accepts ***α*** as name. Please Review.

asmeurer commented on sympy/core/function.py:
```diff
@@ -707,6 +707,11 @@ class UndefinedFunction(FunctionClass):
     """
     def __new__(mcl, name, **kwargs):
+        if PY3:
+            name=str(name)
```
I'd say this line should just be `pass`.

AadityaNair committed:
Optimised code based on suggestions.

No seperate check for python3 was required.

AadityaNair commented:
@asmeurer. Updated PR according to your suggestions. Please Review.

AadityaNair commented on sympy/core/function.py:
```diff
@@ -707,6 +707,9 @@ class UndefinedFunction(FunctionClass):
     The (meta)class of undefined functions.
     """
     def __new__(mcl, name, **kwargs):
+        if not PY3 and isinstance(name, unicode):
```
Leaves PY3 alone

toolforger commented on sympy/core/tests/test_function.py:
```diff
@@ -23,6 +23,25 @@ def test_f_expand_complex():
+    Function('aαβ')
+    Function(u('aαβ'))
+    Function('aαβ') == Function(u('aαβ'))
+
```
The purpose of the first three tests I understand (though it's a bit dubious to label them as "test Greek", "non-ASCII characters" is actually required).
The purpose of the six tests after that I do not understand. What kinds of potential errors do they catch? Can this be documented?

AadityaNair replied:
The thing was that I tried around three implementations to solve this problem.
* The first one solved the case for a single character length function names. But if you added multiple greek characters., they failed.
* Similarly a combination of ASCII and non-ASCII failed my second implementation.

In order to document all these I added them in a test case.
They are basically special cases.

I will work on adding the documentation for them.

AadityaNair committed:
Added documentation in tests.

AadityaNair commented:
@toolforger Added some docs. Please review.

toolforger commented:
Yup, that improves the docs.
And I wasn't aware that you actually had had error situations with these examples; that's indeed a strong argument for including them as test cases.

Still, I find it strange that you had the failures you describe.
Multicharacter UTF-8 isn't substantially different from single-character UTF-8, and from the perspective of UTF-8, 7-bit ASCII characters are just normal UTF-8 characters.
What happened there?

czgdp1807 commented:
Closing in reference to dropping of `Python 2` support. Thanks for the contributions.
